Added MPI functionality

seanr7 opened the pull request.

seanr7 committed:
Changed determinant and integral indexing to start at 0

seanr7 committed:
Changed all orbital indexing to start at 0 and added appropriate comments

seanr7 committed:
Ran black

seanr7 committed:
i hope this works

seanr7 committed:
small formatting edit

seanr7 committed:
ran black

seanr7 committed:
Fixed psi_int indexing

seanr7 committed:
Ran black

seanr7 committed:
added ground state to psi_int

seanr7 committed:
removed comment and added suggestion

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Merge branch 'TApplencourt:clean' into makecategories

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Add category C and tests

seanr7 committed:
Merge branch 'makecategories' of https://github.com/seanr7/QuantumEnvelope into makecategories

seanr7 committed:
Merge branch 'TApplencourt:clean' into makecategories

seanr7 committed:
Merge pull request #4 from seanr7/makecategories

Add category C to clean branch

seanr7 committed:
Removed number of ifs based on spin

seanr7 committed:
Forgot to run black

seanr7 committed:
Thomas's reviews

seanr7 committed:
added revision

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Implemented all categories plus tests

seanr7 committed:
Deleted some old functions + simplified category E

seanr7 committed:
Put old functions back because they made tests fail

seanr7 committed:
Changed double_phase

seanr7 committed:
changed do_double

seanr7 committed:
Cleaning + deleted old stuff

seanr7 committed:
Fixed bug where PT2 tests fail

seanr7 committed:
added comment

seanr7 committed:
Some cleaning, changed do_diagonal

seanr7 committed:
H_indices_idx only loops over canonical integrals and calculates value once

seanr7 committed:
Changed/added minor comments

seanr7 committed:
No longer create new instance of Excitation class each time a cetegory function is called

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Merge branch 'TApplencourt:clean' into clean

seanr7 committed:
Added MPI functionality

seanr7 committed:
black

seanr7 committed:
forgot an argument somewhere

seanr7 committed:
added mpi4py requirement

seanr7 committed:
updated serial davidson

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+    V_k = np.c_[V_k, V_guess]
+
+    n_newvecs = dim_S  # No. of vectors added is initial subspace dimension
+    converged = False
```
I don't think that this variable is needed here

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+
+    n_newvecs = dim_S  # No. of vectors added is initial subspace dimension
+    converged = False
+    restart = 0
```
Look like restart is a boolean, please use it as such

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+                    V_k = np.c_[V_k, t_j / norm_tj]  # Append new vector to trial subspace
+                    n_newvecs += 1
+
+        if converged == True:
```
Please just do `if converged`

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+                t_j = np.dot(M_j, r_j)
+                # Orthogonalize new trial vector against previous basis vectors using Modified Gram-Schmidt (MG)S
+                t_j = t_j / np.linalg.norm(t_j)
+                for i in range(V_k.shape[1]):  # Iterate through k basis vectors
```
This code have like 5 layer of indentation, do you mind try to refractor to avoid this?

For example you start by computing all the `converged` for all eig
```
converged= []
for j in range(n_eig):
      # Grab jth eigenpair, each rank computes local portion of jth residual vector
      l_j, y_j = L_k[j], Y_k[:, j]
      r_j = np.dot(W_k, y_j) - l_j * X_k[:, j]
      res = np.linalg.norm(r_j)
      converged.append( res > eps) 
      
If  all(converged):
    return;

//now handle all the non converged
```
This removes one indentation level.

This should make the code more readable

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
             break
-        theta_old = theta[:eig]
+
+        dim_S += n_newvecs  # Update dimension of trial subspace
```
Look like we don't need to do the addition if `n_newvecs == 0` , should we do this in the `if q<= dims banch` ?

seanr7 replied:
It does get used in the check for that branch.  I could do

```
if q - n_newvecs <= dim_S:  # Collapose trial basis
                self.print_master("q <= dim_S: {dim_S}, restarting Davidson's")
                dim_S, n_newvecs, V_ik, W_ik = self.parallel_restart(
                    dim_S, n_eig, n_newvecs, X_ik, V_ik, W_ik
                )
                restart = True 
                dim_S += n_newvecs
```

seanr7 replied:
Actually, I think this need to be outside, for the cases where we don't enter the restart branches

TApplencourt commented on qe/drivers.py:
```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+
     else:
-        raise NotImplementedError("Not converged")
+        print(f"Max number of iterations reached. Returning {n_eig} lowest eigenvalue estimates")
```
We should still raise, no point in continuing if we don't converge. The `max number of iterations` is here not to waste computation time on non converging hamiltonian 
Do you are @kgasperich ?

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        number_of_dets, res = divmod(len(psi_i), size)
+        # Number of determinants sent to each worker; First `res` processes sent one extra
+        sendcounts = np.array(
+            [number_of_dets + 1 if i < res else number_of_dets for i in range(size)]
```
please put the `dtype` in all the the np.array. This avoid nasty bugs :) (and will made the porting to C++ easier)

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+    comm.Scatterv([indices, sendcounts, displacement, MPI.INT], local_indices, root=0)
+
+    # Create new list of local determinants on each node
+    psi_local = []
+    for I in local_indices:
+        psi_local.append(psi_i[I])
```
```suggestion
    psi_local = [psi_i[i] for for i in local_indices]
```

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+    ):
+        self.d_one_e_integral = d_one_e_integral
+        self.E0 = E0
+        if comm is None:  # Non-MPI jobs
+            self.comm = None
+            self.psi_i = psi_i
+            self.psi_j = psi_j
+        else:  # For MPI jobs
```
I think now that we have only a mpi Davison, the code will break non the less, you can just remove this if. We need to have mpi to run the code

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        # One-electron part
+        for I, det_I in enumerate(self.psi_i):
+            for J, det_J in enumerate(self.psi_j):
+                try:
+                    W_ik[I, :] += (
+                        self.Hamiltonian_one_electron_driver.H_ij(det_I, det_J) * V_k[J, :]
+                    )  # Update row I of W_ik
+                except IndexError:  # If V_k is a single vector.
+                    W_ik[I] += (
+                        self.Hamiltonian_one_electron_driver.H_ij(det_I, det_J) * V_k[J]
+                    )  # Update row I of W_ik
```
This is ugly (you have 2 code path for the vector, matrix case), the trick if you want to support matrix and vector as an input argument is to transform the vector into a matrix as the beginning of the function

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        return W_ik
+
+    def explicit_1e_matrix_product(self, V_k):
+        H = self.H
+
+        return np.dot(H, V_k)
```
```suggestion
        return np.dot(self.H, V_k)
```

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+            return self.H
+        else:
+            H_i = self.H  # Each process builds local Hamiltonian
+            recvcounts = None
```
Useless, please remove

seanr7 committed:
Trying to fix workflow error

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+            H_i = self.H  # Each process builds local Hamiltonian
+            recvcounts = None
+            recvcounts = np.array(
+                self.comm.gather(self.local_size * self.full_size, root=0)
```
please make a choice between `Gather` and `gather`

seanr7 replied:
I think here, the `gather` is okay, since I'm gathering integers into an array. This is what the mpi4py documentation uses in a case like this.

TApplencourt replied:
It will be slower (because of the pickling) and make it confusing for the reader (why sometimes capital or sometimes not?). I have no doubt that the code is valid :) 

I will be more consistent and efficient if first create a `revcount` np.array, and pass it at the `Gather`. It will also make the size and type more explicit, which will ease the porting to C++.

seanr7 replied:
Makes sense! Now it's consistent. :]

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+                L_k = np.zeros(n_eig, dtype="float")
+                Y_k = np.zeros((dim_S, n_eig), dtype="float")
+            # Broadcast smallest n_eig estimates to all ranks to compute residuals
+            self.comm.Bcast([L_k, MPI.DOUBLE], root=0)
```
please use a global variable call MPI_MASTER_RANK to avoid using magic number `root=0`

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+                    print(f"||r_k||: {res}")
+                if res > eps:  # If ||r_j|| > eps, jth eigenpair hasn't converged
+                    converged = False
+                    if self.rank == 0:
+                        print(f"Eigenvalue {j}: not converged, preconditioning next trial vector")
```
To simplify the code, create a function call `print_master` who do this `if rank== MASTER_RANK
print)

seanr7 committed:
revert

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        )  # Build diagonal preconditioner
+        return np.dot(M_k, r_ik)
+
+    def distributed_davidson(self, H_i, n_eig, n_guess, eps, max_iter, q, driven_by="explicit"):
+        """Davidson's method implemented in parallel. The Hamiltonian
```
We can have only one davidson, don't need to save the other one

seanr7 replied:
Done!

TApplencourt reviewed (changes requested):
Can you all add some tests for the Davidson?
Aka comparison Davidson versus `np.eig`.

seanr7 committed:
resolved changes to PR

seanr7 committed:
Merge branch 'clean' of https://github.com/seanr7/QuantumEnvelope into clean

TApplencourt commented on qe/drivers.py:
```diff
@@ -1884,37 +1854,35 @@ def dispatch_psi(comm, psi_i):
         displacement = np.array(
-            [sum(sendcounts[:i]) for i in range(size)]
-        )  # Entry i of displacement vector is the index of the first determinant process i receives
+            [sum(sendcounts[:i]) for i in range(world_size)], dtype="i"
```
Maybe you can use np.sum + axis (https://numpy.org/doc/stable/reference/generated/numpy.sum.html#numpy-sum)

seanr7 replied:
Can you elaborate ?

TApplencourt replied:
`numpy.sum(sendcounts, axis=2)` may be what you are doing manually

seanr7 replied:
I think this sums along a particular array dimension, I'm summing up to a particular index.

perarnau replied:
I think you're looking for `accumulate`

TApplencourt commented on qe/drivers.py:
```diff
@@ -1934,28 +1902,24 @@ def __init__(
+        if self.world_size == 1:  # One process running, use entire wavefunction
+            self.psi_i = psi_j
+        else:  # Else, dispatch chunks of the wavefunction to all processes
             self.psi_i = psi_i
```
Why do this check here?
In theory, when using 1 rank, we still have `chunk` , just one :) 
I would prefer to keep this kind of logic out-side of the function if needed

TApplencourt commented on qe/drivers.py:
```diff
@@ -2014,29 +1977,23 @@ def implicit_1e_matrix_product(self, V_k):
             k = V_k.shape[1]
         except IndexError:  # Handle case when V is a vector
+            V_k = V_k.reshape(len(V_k), 1)
             k = 1
```
Far better! 
Maybe the best is
```
if V_k.ndim == 1
    V_k = V_k.reshape(len(V_k), 1)

k = V_k.shape[1]
```
```

seanr7 replied:
Ah yes! Thanks, I didn't know of this `ndim` command

TApplencourt commented on qe/drivers.py:
```diff
@@ -2190,10 +2137,10 @@ def __init__(
+        if self.world_size == 1:  # One process running, use entire wavefunction
             self.psi_i = psi_j
         else:  # Else, dispatch chunks of the wavefunction to all processes
             self.psi_i = dispatch_psi(comm, psi_j)
```
I think we should always dispatch regardless of the number of processes.   The master thread should still do some work, and not only do the dispatch.
 But maybe we can discuss it, MPI expert may have more insight

seanr7 replied:
I think you're right as well. I also need to figure out where it makes the most sense to do the dispatching. Right now, I have it in this class by default. But it should probably be done when we load the wavefunction I'm thinking?

TApplencourt commented on qe/drivers.py:
```diff
@@ -2453,7 +2408,7 @@ def distributed_davidson(self, H_i, n_eig, n_guess, eps, max_iter, q, driven_by=
             # TODO: Don't need to reduce full matrix at each step. Just new columns/rows
             # Compute rank computes partial update to the projected Hamiltonian S_k
             if (k == 1) or (
```
I think initializating `restart` to True, may remove the check for `k==1`.

seanr7 committed:
more updates, less indentation

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1732,683 @@ def selection_step(
+                f"Distribution of work: {self.distribution}, Local problem size: {self.local_size}"
+            )
+        # Compute offsets (start of the local section) for all nodes
+        self.offsets = [0] + list(accumulate(self.distribution))
+        del self.offsets[-1]
```
```suggestion
        self.offsets = [0] + list(accumulate(self.distribution[:-1]))
```
maybe a  little cleaner

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1732,683 @@ def selection_step(
+        # For easier reference, full problem size
+        self.full_size = problem_size
+        # Set local problem size
+        floor = self.full_size // self.world_size
+        ceiling = floor + 1
+        remainder = self.full_size % self.world_size
```
Please use 
divmod (https://www.programiz.com/python-programming/methods/built-in/divmod) 

floor, remainder = divmod(self.full_size, self.world_size)

TApplencourt commented on qe/drivers.py:
```diff
@@ -1707,3 +1732,683 @@ def selection_step(
+            # Orthogonalize next vector against previous ones in restart basis
+            v_inew, norm_vnew = self.mgs(V_ik[:, :j], V_inew[:, j])
+            V_ik = np.c_[V_ik, v_inew]  # Update basis
+        n_newvecs = dim_S
+
+        return dim_S, n_newvecs, V_ik, W_ik
```
If `n_newvecs` always egual to `dim_S` do we need to return both of them?

seanr7 replied:
No, `dim_S` tracks the overall subspace dimension,`n_newvecs` is the no. of trial vectors added at each iteration. So, the function doesn't need to return this. I can simplify it a little and set it outside of the function

seanr7 committed:
Lots of refactoring. Removed some unused classes + created a Hamiltonian_generator class for each internal wavefunction. Changed powerplant class to take an instance of Davidson class, and changed tests accordingly

kgasperich committed:
testing

kgasperich committed:
testing

kgasperich committed:
testing

kgasperich commented:
https://github.com/seanr7/QuantumEnvelope/pull/5
this fixes the workflow issue

seanr7 committed:
Merge pull request #5 from TApplencourt/mpi4py-ci

install mpi in workflow

seanr7 committed:
wrong type in some class

seanr7 committed:
small changes to class parameters

seanr7 committed:
pt2 selection now in parallel

seanr7 committed:
adding test script

seanr7 committed:
removed some unnecessary lines

seanr7 committed:
Going to push to clean

seanr7 committed:
fixed a doctest

seanr7 committed:
Updated main.py file with new Hamiltonian class. Note: Need to make some changes in the test file. So its okay if these things fail right now

seanr7 committed:
updated tests for new Hamiltonian

seanr7 committed:
no more aliasing in Allreduce buffers

Commit:
issue with buffer dtype

seanr7 committed:
updated test scripts

Commit:
more scripts,plus Davidson's now prints residuals on master rank to track convergence

seanr7 committed:
Switched from try/except to if statements in category functions

seanr7 committed:
Distributed some work of the PT2 selection

seanr7 committed:
update on the fly/cached matrix products. removed dispatch_psi function (not in use)

seanr7 committed:
updated script for a full CIPSI test

seanr7 committed:
Delete test_script for merge

seanr7 committed:
Delete sparsity_plots for merge

seanr7 committed:
Parallel pt2

Adding changes for the parallel PT2 implementation. Assumes psi_external can be stored, but is a working solution for now.

Commit:
Merge branch 'clean' of https://github.com/seanr7/QuantumEnvelope into clean

TApplencourt commented:
Let's merge this! 
Thanks again!
